File: 18. 4Sum/Fixed-size/4sum.cpp
#include<iostream>
#define NOE 4
using namespace std;

int result_number = 0;
int number_of_elements = 4;

class Result{
	int values[NOE];
	int number_of_values;
	bool not_null;
	
	public:
		void print_result(){
			for (int i = 0; i < number_of_values; i++){
				cout << values[i];
				
				if (i < (number_of_values - 1)){
					cout << "\t";
				}
			}
			cout << "\n";
		}
		
		Result (){
			not_null = false;
			number_of_values = 0;
		}
		
		Result (int* list, int len_list){
			not_null = true;
			number_of_values = len_list;
			
			for (int i = 0; i < len_list; i++){
				values[i] = list[i];
			}
		}
};

bool not_equal_list (int* list, int len_list){
	bool out = true;
	
	for (int i = 0; i < len_list; i++){
		for (int j = i; j < len_list; j++){
			if (i != j){
				if (list[i] == list[j]){
					out = false;
				}
			}
		}
	}
	
	return (out);
}

int factorial(int number){
	int out = 1;
	
	for (int i = number; i > 0; i--){
		out *= i;
	}
	
	return (out);
}

void sum_of_four (int* list, int len_list, int target, Result* out){
	int result_cursor = 0;
	
	for (int a = 0; a < len_list; a++){
		for (int b = 0; b < len_list; b++){
			if (b > a){
				for (int c = 0; c < len_list; c++){
					if (c > b){
						for (int d = 0; d < len_list; d++){
							if (d > c){
								int place_equal_list[] = {a, b, c, d};
								int place_equal_list_len = sizeof(place_equal_list)/sizeof(place_equal_list[0]);
								
								if (not_equal_list(place_equal_list, place_equal_list_len)){
									int aa = list[a];
									int bb = list[b];
									int cc = list[c];
									int dd = list[d];
									
									int value_equal_list[] = {aa, bb, cc, dd};
									int value_equal_list_len = sizeof(value_equal_list)/sizeof(value_equal_list[0]);
									
									//if (not_equal_list(value_equal_list, value_equal_list_len)){
									if (true){
										if ((aa + bb + cc + dd) == target){
											int z[] = {list[a], list[b], list[c], list[d]};
											int z_len = sizeof(z)/sizeof(z[0]);
											
											out[result_cursor] = Result(z, NOE);
											result_cursor++;
											result_number = result_cursor;
										}
									}
								}
							}
						}
					}
				}
			}
		}
	}
}

int main (){
	int nums[] = {1,0,-1,0,-2,2};
	int len_nums = sizeof(nums)/sizeof(nums[0]);
	int target = 0;
	
	int sample_space = (factorial(len_nums) / (factorial(len_nums - number_of_elements) * factorial(number_of_elements)));
	
	Result h[sample_space];
	
	sum_of_four(nums, len_nums, target, h);
	
	printf("This is the handy work of C++\n\n");
	
	for (int i = 0; i < result_number; i++){
		h[i].print_result();
	}
	
	printf("\n\n");
	
	return (0);
}

File: 18. 4Sum/Fixed-size/4sum.rs
const number_of_elements: usize = 4;
static mut result_number: usize = 0;
const sample_size: usize = 15;

#[derive(Copy, Clone)]
struct Result {
	values: [isize; number_of_elements],
	no_of_values: usize,
	not_null: bool,
}

impl Result {
	fn new (list: &[isize]) -> Self{
		let len_list: usize = list.len() as usize;
		let mut arr: [isize; number_of_elements] = [0; number_of_elements];
		
		for i in 0..len_list{
			arr[i] = list[i];
		}
		
		return Result{values: arr, no_of_values: len_list, not_null: true};
	}
	
	fn empty () -> Self{
		let arr: [isize; number_of_elements] = [0; number_of_elements];
		
		return Result{values: arr, no_of_values: 0, not_null: false};
	}
	
	fn print_results(&self) -> (){
		for i in 0..self.no_of_values{
			print!("{}", self.values[i]);
			if i < (self.no_of_values - 1){
				print!("\t");
			}
		}
		print!("\n");
	}
}

fn not_equal_list (list: &[isize]) -> bool{
	let mut out: bool = true;
	let len_list: usize = list.len() as usize;
	for i in 0..len_list{
		for j in i..len_list{
			if i != j{
				if list[i] == list[j]{
					out = false;
				}
			}
		}
	}
	return out;
}

fn factorial (number: isize) -> isize{
	let mut out: isize = 1;
	
	for i in (1..=number).rev(){
		out *= i;
	}
	
	return out;
}

fn sum_of_four(list: &[isize], target: isize, out: &mut[Result]){
	let mut result_cursor: usize = 0;
	let len_list: usize = list.len();
	
	for a in 0..len_list{
		for b in 0..len_list{
			if b > a{
				for c in 0..len_list{
					if c > b{
						for d in 0..len_list{
							if d > c{
								if (not_equal_list(&[a as isize, b as isize, c as isize, d as isize])){
									let aa: isize = list[a];
									let bb: isize = list[b];
									let cc: isize = list[c];
									let dd: isize = list[d];
									
									//if (not_equal_list(&[aa, bb, cc, dd]);
									if true{
										if (aa + bb + cc + dd) == target{
											out[result_cursor] = Result::new(&[aa, bb, cc, dd]);
											result_cursor += 1;
											unsafe{
												result_number = result_cursor;
											}
										}
									}
								}
							}
						}
					}
				}
			}
		}
	}
}

fn main (){
	let nums: &[isize] = &[1,0,-1,0,-2,2];
	let len_nums: isize = nums.len() as isize;
	let target = 0;
	
	let sample_space: isize = (factorial(len_nums) / (factorial(len_nums - number_of_elements as isize) * factorial(number_of_elements as isize)));
	
	if (sample_size < sample_space as usize){
		println!("The sample size of all combinations based of the size of the given array is less than the size allocated to handle the results of the program.\nKindly, within the program code file edit the sample size constant\n\"const sample_size: usize = {};\"\nto\n\"const sample_size: usize = {};\"\nPreferably a value that is greater than {}.\n", sample_size, sample_space, sample_space); 
	} else {
		let mut h: [Result; sample_size] = [Result::empty(); sample_size];
		
		sum_of_four(nums, target, &mut h);
		
		println!("The handy work of Rust:\n\n");
		
		let mut rn: isize;
		
		unsafe{
			rn = result_number as isize;
		}
		
		for i in 0..rn{
			h[i as usize].print_results();
		}
	}
	println!("\n\n");
}
